### PR TITLE
Tests reimplementation for Hardhat migration

### PR DESCRIPTION
**Pull request by NZT48**
* Update some deploy scripts
* Implement tests for following contracts: 
   * AssertionStorage
   * ContentAsset
   * Hub
   * Identity
   * IdentityStorage
   * ParametersStorage
   * Profile
   * ProfileStorage
   * ServiceAgreementStorageV1
   * ServiceAgreementV1
   * ShardingTableStorage
   * Staking
   * WhitelistStorage

### CODE REVIEW

**Review by AnaDjokovic — APPROVED** [review]

File: test/assertionStorage.test.ts [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+  it('Get assertion timestamp, size, triples/chunks number for non-existing assertionId, expect to get 0', async () => {
+    const getTimestampResult = await AssertionStorage.getAssertionTimestamp(nonExistingAssertionId);
+    const getSizeResult = await AssertionStorage.getAssertionSize(nonExistingAssertionId);
+    const getTriplesNumber = await AssertionStorage.getAssertionTriplesNumber(nonExistingAssertionId);
+    const getChunksNumber = await AssertionStorage.getAssertionChunksNumber(nonExistingAssertionId);
+
+    expect(getTimestampResult.toString()).to.equal('0');
+    expect(getSizeResult.toString()).to.equal('0');
+    expect(getTriplesNumber.toString()).to.equal('0');
```
Comment (u-hubar) [inline test/assertionStorage.test.ts:91]:
```suggestion
    expect(getTriplesNumber).to.equal(0);
```

File: test/assertionStorage.test.ts [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    expect(getChunksNumber.toString()).to.equal('0');
+  });
+
+  it('Get the assertion timestamp for valid assertion id, expect to pass', async () => {
+    await AssertionStorage.createAssertion(assertionId, size, triplesNumber, chunksNumber);
+    const getTimestampResult = await AssertionStorage.getAssertionTimestamp(assertionId);
+
+    expect(getTimestampResult.toString()).to.not.equal('0');
```
Comment (u-hubar) [inline test/assertionStorage.test.ts:99]:
```suggestion
    expect(getTimestampResult).to.not.equal(0);
```

File: test/assertionStorage.test.ts [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+  });
+
+  it('Get the assertion size for valid assertion id, expect to pass', async () => {
+    await AssertionStorage.createAssertion(assertionId, size, triplesNumber, chunksNumber);
+    const getSizeResult = await AssertionStorage.getAssertionSize(assertionId);
+
+    expect(getSizeResult.toString()).to.equal(size.toString());
```
Comment (u-hubar) [inline test/assertionStorage.test.ts:106]:
```suggestion
    expect(getSizeResult).to.equal(size);
```

File: test/assertionStorage.test.ts [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    expect(getSizeResult.toString()).to.not.equal('0');
+  });
+
+  it('Get the assertion triple number for valid assertion id, expect to pass', async () => {
+    await AssertionStorage.createAssertion(assertionId, size, triplesNumber, chunksNumber);
+    const getTriplesNumber = await AssertionStorage.getAssertionTriplesNumber(assertionId);
+
+    expect(getTriplesNumber.toString()).to.equal(triplesNumber.toString());
```
Comment (u-hubar) [inline test/assertionStorage.test.ts:114]:
```suggestion
    expect(getTriplesNumber).to.equal(triplesNumber);
```

File: test/assertionStorage.test.ts [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+    expect(getTriplesNumber.toString()).to.not.equal('0');
```
Comment (u-hubar) [inline test/assertionStorage.test.ts:115]:
```suggestion
    expect(getTriplesNumber).to.not.equal(0);
```

File: test/assertionStorage.test.ts [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+  });
+
+  it('Get the assertion chunks number for valid assertion id, expect to pass', async () => {
+    await AssertionStorage.createAssertion(assertionId, size, triplesNumber, chunksNumber);
+    const getChunksNumber = await AssertionStorage.getAssertionChunksNumber(assertionId);
+
+    expect(getChunksNumber.toString()).to.equal(chunksNumber.toString());
```
Comment (u-hubar) [inline test/assertionStorage.test.ts:122]:
```suggestion
    expect(getChunksNumber).to.equal(chunksNumber);
```

File: test/assertionStorage.test.ts [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+    expect(getChunksNumber.toString()).to.not.equal('0');
```
Comment (u-hubar) [inline test/assertionStorage.test.ts:123]:
```suggestion
    expect(getChunksNumber).to.not.equal(0);
```

File: test/contentAsset.test.ts [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+    triplesNumber: 10,
+    chunksNumber: 10,
+    epochsNumber: 5,
+    tokenAmount: 250,
+    scoreFunctionId: 1,
+    immutable_: false,
+  };
+
+  async function createAsset() {
+    await Token.mint(accounts[0].address, 1000000000000000);
```
Comment (u-hubar) [inline test/contentAsset.test.ts:38]: Would be cool to use ethers here to make amount of token more readable

File: test/contentAsset.test.ts [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+    assetInputStruct.tokenAmount = 0;
+    await expect(ContentAsset.createAsset(assetInputStruct)).to.be.revertedWithCustomError(
+      ServiceAgreementV1,
+      'ZeroTokenAmount',
+    );
+    assetInputStruct.tokenAmount = 250;
+  });
+
+  it('Create an asset, expect asset created', async () => {
+    await Token.mint(accounts[0].address, 1000000000000000);
```
Comment (u-hubar) [inline test/contentAsset.test.ts:101]: Use ethers?

File: test/contentAsset.test.ts [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+    await Token.increaseAllowance(ServiceAgreementV1.address, assetInputStruct.tokenAmount);
+
+    await expect(ContentAsset.createAsset(assetInputStruct)).to.emit(ContentAsset, 'AssetMinted');
+  });
+
+  it('Get an non existing asset, expect 0 returned', async () => {
```
Comment (u-hubar) [inline test/contentAsset.test.ts:107]: Not accurate description I would say, judging by the test itself

File: test/contentAsset.test.ts [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+    expect(await ContentAssetStorage.getAssertionIds(nonExistingTokenId)).to.deep.equal([]);
+  });
+
+  it('Get an existing asset, expect asset returned', async () => {
```
Comment (u-hubar) [inline test/contentAsset.test.ts:111]: Also here, we're getting basically root hashes of the asset states (assertion identifiers), not really asset itself

File: test/contentAsset.test.ts [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+      'ZeroTokenAmount',
+    );
+    assetInputStruct.tokenAmount = 250;
+  });
+
+  it('Create an asset, expect asset created', async () => {
+    await Token.mint(accounts[0].address, 1000000000000000);
+    await Token.increaseAllowance(ServiceAgreementV1.address, assetInputStruct.tokenAmount);
+
+    await expect(ContentAsset.createAsset(assetInputStruct)).to.emit(ContentAsset, 'AssetMinted');
```
Comment (u-hubar) [inline test/contentAsset.test.ts:104]: Maybe it's not necessary, but we can add args of the event here as well

File: test/staking.test.ts [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+
+    const nodeId1 = '0x07f38512786964d9e70453371e7c98975d284100d44bd68dab67fe00b525cb66';
+    await Profile.createProfile(accounts[0].address, nodeId1, 'Token', 'TKN');
+
+    await Staking['addStake(address,uint72,uint96)'](accounts[0].address, identityId1, 1000000000);
+    expect(await StakingStorage.totalStakes(identityId1)).to.equal(1000000000, 'Total amount of stake is not set');
+  });
+
+  it('Add reward; expect that total stake is increased', async function () {
+    await Token.mint(ServiceAgreementStorageV1.address, 1000000000000000);
```
Comment (u-hubar) [inline test/staking.test.ts:141]: Use ethers for token amount readability?

File: test/shardingTableStorage.test.ts [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+    expect(getNodeResult.identityId.toNumber()).to.equal(profiles[1]);
+    expect(getNodeResult.prevIdentityId.toNumber()).to.equal(profiles[0]);
+    expect(getNodeResult.nextIdentityId.toNumber()).to.equal(profiles[2]);
+
+    // add 1 more node
+    const adminWallet3 = await Profile.connect(accounts[3]);
+    const newProfile = await adminWallet3.createProfile(accounts[0].address, nodeId4, 'Token4', 'TKN4');
+
+    const receipt1 = await newProfile.wait();
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
```
Comment (u-hubar) [inline test/shardingTableStorage.test.ts:187]: It's possible to disable for the whole file on the top

File: test/shardingTableStorage.test.ts [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+    // @ts-ignore
+    const newIdentityId = receipt1.events[3].args.identityId.toNumber();
```
Comment (u-hubar) [inline test/shardingTableStorage.test.ts:189]: Not sure about this, but I guess it should work without toNumber(), using BigNumber object

File: test/shardingTableStorage.test.ts [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+    expect(getLinkedNodeValues.prevIdentityId).to.equal(profiles[1]); // 2
+    expect(getLinkedNodeValues.nextIdentityId).to.equal(profiles[2]); // 3
+
+    const getValuesForNext2Nodes = await ShardingTableStorage.getMultipleNodes(profiles[1], 2);
+
+    // NODE[2] has 2 linked nodes:
+    // FIRST: previous identityId 1 and nextIdentity id 4
+    // SECOND: previous identityId 2 and nextIdentity id 3
+    expect(getValuesForNext2Nodes.length).to.equal(2);
+    expect(getValuesForNext2Nodes[0].identityId.toNumber()).to.equal(profiles[1]);
```
Comment (u-hubar) [inline test/shardingTableStorage.test.ts:206]:
```suggestion
    expect(getValuesForNext2Nodes[0].identityId).to.equal(profiles[1]);
```

Comment (u-hubar) [reply test/shardingTableStorage.test.ts:206]: Also same comment for everything below :)

File: test/identityStorage.test.ts [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+  let Identity: Identity;
+  let Hub: Hub;
+  let IdentityStorage: IdentityStorage;
+  let operationalKey: string,
+    adminKey: string,
+    newAdminKeyBytes32: string,
+    operationalKeyBytes32: string,
+    adminKeyBytes32: string,
+    newOperationalKeyBytes32: string;
+  const ADMIN_KEY = 1;
```
Comment (u-hubar) [inline test/identityStorage.test.ts:27]: Maybe makes sense to move it to constants file?

File: test/shardingTableStorage.test.ts [hunk 18]
```diff
@@ -0,0 +1,213 @@
...
+    expect(getValuesForNext2Nodes[0].prevIdentityId.toNumber()).to.equal(profiles[0]);
```
Comment (NZT48) [inline test/shardingTableStorage.test.ts:207]:
```suggestion
    expect(getValuesForNext2Nodes[0].prevIdentityId).to.equal(profiles[0]);
```

File: test/shardingTableStorage.test.ts [hunk 19]
```diff
@@ -0,0 +1,213 @@
...
+    expect(getValuesForNext2Nodes[0].nextIdentityId.toNumber()).to.equal(newIdentityId);
```
Comment (NZT48) [inline test/shardingTableStorage.test.ts:208]:
```suggestion
    expect(getValuesForNext2Nodes[0].nextIdentityId).to.equal(newIdentityId);
```

File: test/shardingTableStorage.test.ts [hunk 20]
```diff
@@ -0,0 +1,213 @@
...
+    expect(getValuesForNext2Nodes[1].identityId.toNumber()).to.equal(newIdentityId);
```
Comment (NZT48) [inline test/shardingTableStorage.test.ts:209]:
```suggestion
    expect(getValuesForNext2Nodes[1].identityId).to.equal(newIdentityId);
```

File: test/shardingTableStorage.test.ts [hunk 21]
```diff
@@ -0,0 +1,213 @@
...
+    expect(getValuesForNext2Nodes[1].prevIdentityId.toNumber()).to.equal(profiles[1]);
```
Comment (NZT48) [inline test/shardingTableStorage.test.ts:210]:
```suggestion
    expect(getValuesForNext2Nodes[1].prevIdentityId).to.equal(profiles[1]);
```

File: test/shardingTableStorage.test.ts [hunk 22]
```diff
@@ -0,0 +1,213 @@
...
+    expect(getValuesForNext2Nodes[1].nextIdentityId.toNumber()).to.equal(profiles[2]);
```
Comment (NZT48) [inline test/shardingTableStorage.test.ts:211]:
```suggestion
    expect(getValuesForNext2Nodes[1].nextIdentityId).to.equal(profiles[2]);
```

**Pull request merged**